File: src/main/java/tn/esprit/services/GeminiService.java
package tn.esprit.services;

import okhttp3.*;
import org.json.JSONArray;
import org.json.JSONObject;
import tn.esprit.utils.myDatabase;

import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

public class GeminiService {
    private static final String API_KEY = "AIzaSyCIYfKgYYYMmuUzvKlAKi8byyebAri5cgQ"; // Remplace par ta vraie clé API
    private static final String API_URL = "https://generativelanguage.googleapis.com/v1/models/gemini-1.5-pro:generateContent?key=" + API_KEY;

    private final OkHttpClient client;
    private Connection cnx ;

    public GeminiService() {
        this.client = new OkHttpClient();
        cnx = myDatabase.getInstance().getConnection();
    }

    public String getResponse(String userInput) {
        if (userInput.toLowerCase().contains("événements disponibles") || userInput.toLowerCase().contains("quel événement") || userInput.toLowerCase().contains("événement") || userInput.toLowerCase().contains("événements")) {
            return getUpcomingEvents(); // 🔹 Récupération des événements depuis la BDD
        }

        if (userInput.toLowerCase().contains("espaces disponibles") || userInput.toLowerCase().contains("quel espace") || userInput.toLowerCase().contains("espace") || userInput.toLowerCase().contains("espaces")) {
            return getAvailableSpaces(); // 🔹 Vérification des espaces disponibles
        }

        return askGemini(userInput); // 🔹 Si la question ne concerne pas la BDD, on utilise l'API Gemini
    }

    // 🔹 Méthode pour récupérer les événements disponibles depuis `lammaInteg`
    private String getUpcomingEvents() {
        StringBuilder response = new StringBuilder("🎟️ Voici les événements disponibles :\n");

        try {
            String query = "SELECT nomEvent, prix, date, nomEspace FROM event ";
            PreparedStatement stmt = cnx.prepareStatement(query);
            ResultSet rs = stmt.executeQuery();

            while (rs.next()) {
                response.append("🎉 **").append(rs.getString("nomEvent")).append("**\n")
                        .append("📅 Date : ").append(rs.getString("date")).append("\n")
                        .append("📍 Lieu : ").append(rs.getString("nomEspace")).append("\n")
                        .append("💰 Prix : ").append(rs.getInt("prix")).append(" TND\n\n");
            }

            if (response.toString().equals("🎟️ Voici les événements disponibles :\n")) {
                return "Aucun événement n'est prévu pour le moment. Revenez plus tard !";
            }

            return response.toString();
        } catch (Exception e) {
            return "Erreur lors de la récupération des événements : " + e.getMessage();
        }
    }

    // 🔹 Récupération des espaces disponibles
    private String getAvailableSpaces() {
        StringBuilder response = new StringBuilder("📍 Voici les espaces disponibles :\n");

        try {
            String query = "SELECT nomEspace, adresse, capacite, disponibilite FROM espace WHERE disponibilite = 'Disponible'";
            PreparedStatement stmt = cnx.prepareStatement(query);
            ResultSet rs = stmt.executeQuery();

            while (rs.next()) {
                response.append("🏢 **").append(rs.getString("nomEspace")).append("**\n")
                        .append("📍 Adresse : ").append(rs.getString("adresse")).append("\n")
                        .append("👥 Capacité : ").append(rs.getInt("capacite")).append(" personnes\n\n");
            }

            if (response.toString().equals("📍 Voici les espaces disponibles :\n")) {
                return "Aucun espace n'est disponible en ce moment.";
            }

            return response.toString();
        } catch (Exception e) {
            return "Erreur lors de la récupération des espaces : " + e.getMessage();
        }
    }

    // 🔹 Méthode pour interagir avec l'API Gemini si nécessaire
    private String askGemini(String userInput) {
        try {
            String promptContext = "Tu es un chatbot d'assistance pour une plateforme de gestion d'événements en ligne. " +
                    "Tu aides les utilisateurs à trouver des événements, consulter les espaces disponibles, réserver des billets " +
                    "et obtenir des informations sur les tarifs et les disponibilités. Réponds de manière claire et interactive.";

            String fullPrompt = promptContext + "\n\nUtilisateur : " + userInput;

            JSONObject requestBody = new JSONObject();
            JSONArray contentsArray = new JSONArray();
            JSONObject contentObject = new JSONObject();
            contentObject.put("role", "user");
            contentObject.put("parts", new JSONArray().put(new JSONObject().put("text", fullPrompt)));
            contentsArray.put(contentObject);
            requestBody.put("contents", contentsArray);

            Request request = new Request.Builder()
                    .url(API_URL)
                    .post(RequestBody.create(requestBody.toString(), MediaType.get("application/json; charset=utf-8")))
                    .build();

            Response response = client.newCall(request).execute();

            if (!response.isSuccessful()) {
                return "Erreur API " + response.code() + " : " + response.message();
            }

            String responseBody = response.body().string();
            JSONObject jsonResponse = new JSONObject(responseBody);

            if (jsonResponse.has("error")) {
                return "Erreur API : " + jsonResponse.getJSONObject("error").toString();
            }

            return jsonResponse.optJSONArray("candidates")
                    .optJSONObject(0)
                    .optJSONObject("content")
                    .optJSONArray("parts")
                    .optJSONObject(0)
                    .optString("text", "Erreur : Aucun texte retourné.");
        } catch (IOException e) {
            return "Erreur de connexion : " + e.getMessage();
        }
    }


}

File: src/Service/GeminiService.php
<?php
// src/Service/GeminiService.php
namespace App\Service;

use Doctrine\DBAL\Connection;
use Symfony\Contracts\HttpClient\HttpClientInterface;

class GeminiService
{
    private string $apiKey = 'AIzaSyCIYfKgYYYMmuUzvKlAKi8byyebAri5cgQ';
    private string $apiUrl = 'https://generativelanguage.googleapis.com/v1/models/gemini-1.5-pro:generateContent?key=';

    public function __construct(private Connection $conn, private HttpClientInterface $httpClient) {}

    public function getResponse(string $prompt): string
{
    $userInput = strtolower($prompt);

    if ($this->containsAny($userInput, ['événement', 'événements', 'event', 'disponible', 'quels événements', 'liste des événements'])) {
        return $this->getEvents();
    }

    if ($this->containsAny($userInput, ['espace', 'espaces', 'lieux', 'salles', 'disponibles', 'réserver un espace'])) {
        return $this->getEspaces();
    }

    if ($this->containsAny($userInput, ['produit', 'produits', 'catalogue', 'stock', 'articles', 'objets'])) {
        return $this->getProduits();
    }

    return $this->askGemini($prompt);
}

private function containsAny(string $text, array $keywords): bool
{
    foreach ($keywords as $word) {
        if (str_contains($text, $word)) {
            return true;
        }
    }
    return false;
}

    private function getEvents(): string
    {
        $res = $this->conn->executeQuery('SELECT nomEvent, prix, date, nomEspace FROM event')->fetchAllAssociative();
        if (!$res) return 'Aucun événement trouvé.';
        $txt = "🎟️ Événements disponibles :\n";
        foreach ($res as $e) {
            $txt .= "🎉 {$e['nomEvent']} | 📅 {$e['date']} | 📍 {$e['nomEspace']} | 💰 {$e['prix']} TND\n";
        }
        return $txt;
    }

    private function getEspaces(): string
    {
        $res = $this->conn->executeQuery("SELECT nomEspace, adresse, capacite FROM espace WHERE disponibilite = 'Disponible'")->fetchAllAssociative();
        if (!$res) return 'Aucun espace disponible.';
        $txt = "📍 Espaces disponibles :\n";
        foreach ($res as $e) {
            $txt .= "🏢 {$e['nomEspace']} | 📍 {$e['adresse']} | 👥 Capacité : {$e['capacite']} pers.\n";
        }
        return $txt;
    }

    private function getProduits(): string
    {
        $res = $this->conn->executeQuery("SELECT nomProduit, prixProduit, quantite FROM produit WHERE quantite > 0")->fetchAllAssociative();
        if (!$res) return 'Aucun produit en stock.';
        $txt = "🛒 Produits disponibles :\n";
        foreach ($res as $p) {
            $txt .= "📦 {$p['nomProduit']} | 💰 {$p['prixProduit']} TND | 📦 Stock : {$p['quantite']}\n";
        }
        return $txt;
    }

    private function askGemini(string $prompt): string
    {
        $fullPrompt = "Tu es un assistant pour une plateforme de gestion d'événements. Réponds clairement.\n\nUtilisateur : $prompt";
        $json = [
            "contents" => [[
                "role" => "user",
                "parts" => [["text" => $fullPrompt]]
            ]]
        ];

        $response = $this->httpClient->request('POST', $this->apiUrl . $this->apiKey, [
            'json' => $json
        ]);

        $data = $response->toArray(false);
        return $data['candidates'][0]['content']['parts'][0]['text'] ?? '🤖 Je n’ai pas compris. Réessaie.';
    }
}
?>